clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -268,6 +280,13 @@ public void poll(long timeout) throws IOException {
         currentTimeNanos = endSelect;
         this.sensors.selectTime.record(endSelect - startSelect, time.milliseconds());
 
+        // Perform setup for instantly connected channels since they won't be included in selectedKeys()
+        if (connectableChannels.size() > 0)
+            for (KafkaChannel channel : connectableChannels) {
+                sensors.maybeRegisterConnectionMetrics(channel.id());
```

Comment:
@llowrey, it would also be interesting to know the environment where you can reproduce this. I tried writing a test that connects 1 million times to a Java EchoServer and `connect` always returns false.